File: 566-reshape-matrix/reshape_matrix.go
package main

import "fmt"

func matrixReshape(mat [][]int, r int, c int) [][]int {
	matrixRows := len(mat)
	matrixCols := len(mat[0])

	if matrixRows*matrixCols != r*c {
		return mat
	}

	col := 0
	row := 0
	resultMatrix := make([][]int, r)
	for i := 0; i < r; i++ {
		resultMatrix[i] = make([]int, c)
	}

	for i := 0; i < len(mat); i++ {
		for j := 0; j < len(mat[0]); j++ {
			if col == c {
				row++
				col = 0
			}

			resultMatrix[row][col] = mat[i][j]
			col++
		}

	}

	return resultMatrix
}

func main() {
	matrix_1 := [][]int{
		{1, 2, 3, 4},
		{5, 6, 7, 8},
		{9, 10, 11, 12},
	}
	result := matrixReshape(matrix_1, 2, 6)
	fmt.Println(result)
}

File: 0566-reshape-matrix/reshapeMatrix.js
/**
 * @param {number[][]} mat
 * @param {number} r
 * @param {number} c
 * @return {number[][]}
 */
const matrixReshape = function(mat, r, c) { 
  const matrixRows = mat.length; 
  const matrixCols = mat[0].length; 

  if (matrixRows * matrixCols !== r * c) {
    return mat; 
  } 

  let col = 0; 
  let row = 0; 

  const resultArray = (new Array(r)).fill().map(() => Array(c).fill(0))

  for (let i = 0; i < mat.length; i++) {
    for (let j = 0; j < mat[0].length; j++) {
      if (col === c) {
        row++; 
        col = 0; 
      }
      resultArray[row][col] = mat[i][j]; 
      col++; 
    }
  }

  return resultArray; 

};



console.log(matrixReshape([[1, 2], [3, 4]], 1, 4)); // [[1, 2, 3, 4]]
console.log(matrixReshape([[1, 2], [3, 4]], 1, 3)); // [[1, 2], [3, 4]]

console.log(matrixReshape([[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]], 4, 3)); // [[1, 2, 3], [4, 5, 6], [7, 8, 9], [10, 11, 12]]